File: sessions_to_encounters.go
// Note this script should produce the same encounter set as the Python version
// for a given set of input sessions, but may not emit them in the same order.
// I believe this is being caused by Python and Go using different sorting
// algorithms meaning the session list can have more than one valid sorting
// leading to more than one valid encounter list ordering.

package main

import (
	"fmt"
	"bufio"
	"strings"
	"strconv"
	"sort"
	"os"
)

func main() {
	// read sessions from stdin
	sessions := []Session{}	
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		fields := strings.Split(strings.TrimSpace(scanner.Text()), ",")
		node := fields[0]
		start, _ := strconv.Atoi(fields[1])
		end, _ := strconv.Atoi(fields[2])
		location := fields[3]
		sessions = append(sessions, Session{node, start, end, location})
	}

	sort.Sort(ByStart(sessions))
	// key -> location, value -> session list
	cand_encs := make(map[string][]Session)
	contacts := []Contact{}
	imaginary_contact_count := 0
	for _, s := range sessions {
		loc := s.location
		_, ok := cand_encs[loc]
		if !ok {
			cand_encs[loc] = []Session{}
		} else {
			// only sessions with end > this sess start remain
			// candidates.  
			ind := 0
			for _, x := range cand_encs[loc] {
				if x.end > s.start {
					cand_encs[loc][ind] = x
					ind++
				}
			}
			cand_encs[loc] = cand_encs[loc][:ind]
		}
		/* sessions with end < this sess start already filtered.  Any
		/* remaining sessions in cand_encs[loc] are for same loc and
		/* started prior to or at same time as this session by virtue of
		/* sessions being sorted by start time.  Therefore all remianing
		/* sessions in cand_encs[loc] started before this session
		/* started and ended after this session started, implying an
		/* encounter. */
		for _, s2 := range cand_encs[loc] {
			if s.node == s2.node {
				// Can happen under e.g. shuffled sessions.
				imaginary_contact_count++
				continue
			}
			var enc_start, enc_end int
			if s.start > s2.start { 
				enc_start = s.start 
			} else { 
				enc_start = s2.start 
			}
			if s.end < s2.end { 
				enc_end = s.end 
			} else { 
				enc_end = s2.end 
			}
			contacts = append(contacts, Contact{s.node, s2.node,
				enc_start, enc_end, loc})
		}
		cand_encs[loc] = append(cand_encs[loc], s)
	}
	
	// write encounters to stdout
	for _, c := range contacts {
		fmt.Println(strings.Join([]string{
			c.node1,
			c.node2,
			strconv.Itoa(c.start),
			strconv.Itoa(c.end),
			c.location},
			","))
	}
	//fmt.Fprintf(os.Stderr, "Number of imaginary contacts: %d\n", 
	//	imaginary_contact_count)
}

type Session struct {
	node string
	start, end int
	location string
}

type Contact struct {
	node1, node2 string
	start, end int
	location string
}

// sorting interface by session start time.
type ByStart []Session

func (a ByStart) Len() int { return len(a) }
func (a ByStart) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
func (a ByStart) Less(i, j int) bool { return a[i].start < a[j].start }

File: sessions_to_encounters.py
"""Convert sessions to encounters.

If called as main script:

stdin

    A set of comma-separated value lines describing sessions , each of the form
    <node, start, end, location>.

stdout

    A set of comma-separated value lines describing contact events, each of the
    form <node_1, node_2, start, end, location>.

"""

import sys
from collections import defaultdict

def sessions_to_encounters(sessions, start_time=None):
    """Return a set of encounters from a set of sessions.
    
    Parameters

        sessions : a list of 4-tuples of the form <node, start, end, location>,
        describing the times at which individual nodes were present at individual
        locations.

        start_time : if specified, only returns contacts that have
        occurred after the designated start time.  Contacts which
        straddle the designated start time are ltrimmed to commence at the
        designated start time.
    
    Returns

        A list of 5-tuples of the form <node_1, node_2, start, end, location>,
        describing the intervals during which two connected devices encountered.

    """
    # sort by start_time - prereq for implemented  encounter algorithm.
    sessions.sort(key = lambda x: x[1])
    candidate_encounters = defaultdict(list)
    encounters = []
    for sess in sessions:
        ap = sess[3]
        # only sessions with end time > this sess start time remain
        # candidates
        candidate_encounters[ap] = [x for x in \
        candidate_encounters[ap] if x[2] > sess[1]]
        
        # Sessions with end time < start time already filtered.
        # Any remaining sessions in candidate_encounters[ap]
        # are for the same AP and started prior to or at the
        # same time as this session by virtue of sessions being
        # sorted by start time.  Therefore all remaining sessions
        # in candidate_encounters AP started before this session
        # started and ended after this session started, implying
        # an encounter.
        for enc in candidate_encounters[ap]:
            enc_start = max(sess[1], enc[1])
            enc_end = min(sess[2], enc[2])
            mac_1 = sess[0]
            mac_2 = enc[0]
            #assert mac_1 != mac_2, 'Mac should not encounter itself'
            # have to check this, as with e.g. optimistic session adjustments
            # a MAC's sessions can overlap in time even at the same AP.
            if mac_1 != mac_2:
                encounters.append([mac_1, mac_2, enc_start, enc_end, ap])
        
        candidate_encounters[ap].append(sess) 
    # if no start_time has been designated, return now. otherwise,
    # discard encounters before start time and ltrim those which
    # straddle it.
    if not start_time:   
        return encounters
    else:
        # encounter must end after designated start time
        encounters = [enc for enc in encounters if enc[3] > start_time]
        # ltrim encounters if necessary
        new_encounters = []
        for enc in encounters:
            if enc[2] < start_time:
                new_encounters.append([enc[0],enc[1],start_time,enc[3],enc[4]])
            else:
                new_encounters.append(enc)
        return new_encounters


if __name__ == "__main__":
    sessions = []
    for line in sys.stdin:
        # fields
        f = line.strip().split(',')
        sessions.append([f[0], int(f[1]), int(f[2]), f[3]])
    encounters = sessions_to_encounters(sessions)
    for e in encounters:
        print ','.join(map(str, e))
